Source/DafnyTests/DafnyTestCase.cs

```diff
@@ -0,0 +1,417 @@
+
+    private static DirectoryInfo OUTPUT_ROOT = new DirectoryInfo(Directory.GetCurrentDirectory());
+    
+    // TODO-RS: This is an ugly method of locating the project root.
+    // The proper fix is to run entirely out of the output directory,
+    // and the projects are at least partially configured to make that possible,
+    // but it's not quite working yet.
+    private static string DAFNY_ROOT = 
+      OUTPUT_ROOT.Parent.Parent.Parent.Parent.Parent.FullName;
+
+    public static readonly string TEST_ROOT = Path.Combine(DAFNY_ROOT, "Test") + Path.DirectorySeparatorChar;
+    public static readonly string COMP_DIR = Path.Combine(TEST_ROOT, "comp") + Path.DirectorySeparatorChar;
+    public static readonly string OUTPUT_DIR = Path.Combine(TEST_ROOT, "Output") + Path.DirectorySeparatorChar;
+    
+    public static readonly string DAFNY_EXE = Path.Combine(DAFNY_ROOT, "Binaries/dafny");
+
+    // Dafny options with special handling
+    public const string DAFNY_COMPILE_OPTION = "compile";
+    public const string DAFNY_COMPILE_TARGET_OPTION = "compileTarget";
+    public const string DAFNY_NO_VERIFY_OPTION = "noVerify";
+
+    // Absolute file system path to the main Dafny file
+    [YamlIgnore]
+    public readonly string SourcePath;
+    
+    public Dictionary<string, object> DafnyArguments = new Dictionary<string, object>();
+    public List<string> OtherFiles = new List<string>();
+    
+    public Dictionary<string, DafnyTestSpec> CompileTargetOverrides = new Dictionary<string, DafnyTestSpec>();
+    
+    public DafnyTestCase.Expectation Expected;
+
+    public DafnyTestSpec(string sourcePath) {
+      SourcePath = sourcePath;
+    }
+
+    private DafnyTestSpec(string sourcePath, Dictionary<string, object> dafnyArguments, List<string> otherFiles, DafnyTestCase.Expectation expected) {
+      SourcePath = sourcePath;
+      DafnyArguments = dafnyArguments;
+      OtherFiles = otherFiles;
+      Expected = expected;
+    }
+    
+    public IEnumerator<DafnyTestCase> GetEnumerator() {
+      return ResolveCompile()
+        .SelectMany(spec => spec.ExpandArguments())
+        .Select(spec => spec.ResolveExpected().ToTestCase())
+        .GetEnumerator();
+    }
+
+    public string Compile {
+      get {
+        if (DafnyArguments.TryGetValue(DAFNY_COMPILE_OPTION, out var compile)) {
+          return (string) compile;
+        }
+        return null;
+      }
+      set {
+        DafnyArguments[DAFNY_COMPILE_OPTION] = value;
+      }
+    }
+    public string CompileTarget {
+      get {
+        if (DafnyArguments.TryGetValue(DAFNY_COMPILE_TARGET_OPTION, out var compileTarget)) {
+          return (string) compileTarget;
+        }
+        return null;
+      }
+    }
+
+    private IEnumerable<DafnyTestSpec> ResolveCompile() {
+      if (Compile == null) {
+        Compile = "3";
+      }
+      
+      if ("3".Equals(Compile) && CompileTarget == null) {
+        var compileTargets = new[] {"cs", "java", "go", "js"};
```

Comment:
We should probably have a static const of all supported languages somewhere higher so we can reduce the chance of missing this as new compile targets are added.

Reply:
Couldn't agree more, this should just refer directly to the `CompilationTarget` enumeration. I've only got it as a separate list temporarily and already added a TODO here.